### PR TITLE
GEOMESA-1085 Redefine Config Structure

### PR DESCRIPTION
**Pull request by jahhulbert-ccri**
* SFT and Converter config are no longer arrays
* Configs can be concatenated and combined into a single file (e.g. with the maven appending resource transformer) and still be read.

Signed-off-by: Andrew Hulbert <andrew.hulbert@ccri.com>

See this branch for example:
https://github.com/geomesa/gm-data/tree/fcr_gmdata_all

### CODE REVIEW

File: geomesa-tools/conf/application.conf [hunk 1]
```diff
@@ -11,36 +10,32 @@ geomesa = {
...
     }
-    {
-      type-name = "example-json"
+    "example-json" = {
       attributes = [
-        { name = "number", type = "Integer" }
-        { name = "color",  type = "String"  }
-        { name = "weight", type = "Double"  }
-        { name = "geom",   type = "Point"   }
+        {name = "number", type = "Integer"}
```
Comment (jahhulbert-ccri) [inline geomesa-tools/conf/application.conf]: ugh...sorry i think intellij did this to me

File: geomesa-convert/README.md [hunk 2]
```diff
@@ -824,5 +827,29 @@ updated file.
...
         }
-      ]
+      }
     }
+
+## Loading Converters and SFTs at Runtime
+
+If you have defined converters or SFTs in typesafe config you can place them on the classpath or load them with a 
+ConverterConfigProvider or SimpleFeatureTypeProvider via Java SPI loading. By default, classpath and URL providers are 
+provided. Placing a typesafe config file named ``reference.conf`` containing properly formatted converters and SFTs
```
Comment (elahrvivaz) [inline geomesa-convert/README.md]: i think the preferred scheme is for only libraries to use reference.conf, and applications to use application.conf.

File: geomesa-tools/conf/application.conf [hunk 3]
```diff
@@ -1,7 +1,6 @@
-geomesa = {
-  sfts = [
-    {
-      type-name = "example-csv"
```
Comment (elahrvivaz) [inline geomesa-tools/conf/application.conf]: nice, i was going to suggest replacing type name with the config name.

**Comment by elahrvivaz** [comment]
looks good - need to update the docs too, though.

File: geomesa-tools/conf/application.conf (hunk 3 above)
Comment (jahhulbert-ccri) [inline geomesa-tools/conf/application.conf]: interestingly enough i guess the config name could be techincally different from the SFT type name...i think we want to get away from it all together...but you do have the weird case where you provide it as an arg by itself...we could force people to nest it properly in ```geomesa.sfts.<name>``` in that case which might be best.

**Comment by jahhulbert-ccri** [comment]
i'll look through for more doc stuff. i added some. we didn't have much except in the converter config section :) Is there a better place for it? I'll look through the docs that mzim added

**Comment by mdzimmerman** [comment]
The converters are mentioned in a couple of places:

* ``docs/user/commandline_tools.rst``
* ``docs/user/examples.rst``

It's also mentioned in ``docs/developer/modules/geomesa-convert.rst``, but that's basically just a copy of the ``geomesa-convert/README.md`` file, and can be sucked in with a script in ``docs/bin``.

File: geomesa-convert/README.md (hunk 2 above)
Comment (jahhulbert-ccri) [inline geomesa-convert/README.md]: should i change this wording?

File: docs/user/commandline_tools.rst [hunk 4]
```diff
@@ -170,22 +170,26 @@ ingest
 
 Used to convert and ingest data from various file formats as GeoMesa features.
 
-CSV and TSV files   can be ingested either with explicit latitude and longitude columns or with a column of WKT geometries.
-For lat/lon column ingest, the ``SimpleFeatureType`` spec must include an additional geometry attribute beyond the number of 
-columns in the file such as: `*geom:Point`. The file type is inferred from the extension of the file, so ensure that 
-the formatting of the file matches the extension of the file and that the extension is present.
+GeoMesa defines several common converter factories for formats such as delimited text
+(TSV, CSV), fixed width files, json, xml, and avro. New converter factories (e.g. for custom binary formats) can be
```
Comment (mdzimmerman) [inline docs/user/commandline_tools.rst]: (picky) JSON and XML are initialisms and should be written in caps.

File: docs/user/commandline_tools.rst [hunk 5]
```diff
@@ -170,22 +170,26 @@ ingest
...
+registered by packaging a ``reference.conf`` file inside a jar on the classpath or by adding a new converter definition
+to the ``geomesa-tools-$VERSION/conf/application.conf`` file. For example::
 
     # cat $GEOMESA_HOME/conf/application.conf
     geomesa {
-      sfts = [
-        {
-          type-name = "renegades"
+      sfts {
+        renegages = {
```
Comment (mdzimmerman) [inline docs/user/commandline_tools.rst]: Misspelled "renegades"?

File: geomesa-convert/README.md [hunk 6]
```diff
@@ -781,17 +781,21 @@ Sample csv file: ``example.csv``:
 
 The "renegades" SFT and "renegades-csv" converter should be specified in 
 the GeoMesa Tools configuration file (``$GEOMESA_HOME/conf/application.conf``).
+By default, SimpleFeatureTypes (sfts) should be loaded at the path ``geomesa.sfts``
+and conveters should be loaded at the path ``geomesa.converters``. Each converter
```
Comment (mdzimmerman) [inline geomesa-convert/README.md]: s/conveters/converters/. Also (picky) I would style "sfts" as "SFTs".

File: geomesa-tools/README.md [hunk 7]
```diff
@@ -539,9 +542,8 @@ the formatting of the file matches the extension of the file and that the extens
         
     # cat $GEOMESA_HOME/conf/application.conf
     geomesa {
-      sfts = [
-        {
-          type-name = "renegades"
+      sfts {
+        renegages = {
```
Comment (mdzimmerman) [inline geomesa-tools/README.md]: Misspelled "renegades".

**Comment by jahhulbert-ccri** [comment]
oops...reopening

File: geomesa-convert/README.md (hunk 6 above)
Comment (jahhulbert-ccri) [inline geomesa-convert/README.md]: fixed

File: docs/user/commandline_tools.rst (hunk 5 above)
Comment (jahhulbert-ccri) [inline docs/user/commandline_tools.rst]: fixed

File: geomesa-tools/README.md (hunk 7 above)
Comment (jahhulbert-ccri) [inline geomesa-tools/README.md]: fixed

File: geomesa-convert/README.md (hunk 2 above)
Comment (elahrvivaz) [inline geomesa-convert/README.md]: eh, it's ok

File: geomesa-tools/conf/application.conf (hunk 3 above)
Comment (elahrvivaz) [inline geomesa-tools/conf/application.conf]: yeah, i like that idea.

File: geomesa-tools/conf/application.conf [hunk 8]
```diff
@@ -1,46 +1,41 @@
...
       ]
     }
-    {
-      type-name = "example-avro"
+    "example-avro" = {
       attributes = [
-        { name =  "dtg", type = "Date", default = true }
-        { name = "geom", type = "Point", index = "true", srid = 4326, default = true  }
+        {name = "dtg",  vtype = "Date", default = true }
+        {name = "geom", type = "Point", index = "true", srid = 4326, default = true }
```
Comment (elahrvivaz) [inline geomesa-tools/conf/application.conf]: kind of picky, but we should try to be consistent with spaces before/after {} - right now we have them before but not after

Comment (jahhulbert-ccri) [inline geomesa-tools/conf/application.conf]: yeah i'll standardize it

File: geomesa-tools/conf/application.conf (hunk 3 above)
Comment (jahhulbert-ccri) [inline geomesa-tools/conf/application.conf]: cool i'm gonna make that change first

**Comment by jahhulbert-ccri** [comment]
good to go for now...gonna think about some more stuff for command line args while i'm writing tests for converters.

**Comment by jahhulbert-ccri** [comment]
gonna add one more fix.

**Comment by elahrvivaz** [comment]
38bdb73368b6e357e258d8538baf2bd8e0cb8979

**Pull request closed without merging**